apps/labs/posts/numpy-thread-safe-test-suite.md

```diff
@@ -0,0 +1,375 @@
+
+Of course, you don't just have to read up on articles, you can give it a try yourself! Python's latest version, 3.14, comes with the option to install the free-threaded build (which can be denoted as "3.14t").
+
+(It's a fun coincidence that Hollow Knight: Silksong came out around midway through my internship. Get it? *Silk*song, free-_threaded_?)
+
+### The Codebase & Test Suite
+
+With free-threaded Python getting more support, it's a good idea to ensure your code can handle running on multiple threads. One way of doing this is to run the test suite under multiple threads. The test suite already attempts to test the codebase, so if we run the tests in multiple threads, we could see if the codebase plays nice with threads.
+
+**Note:** This approach, while straightforward, will only find a subset of possible issues. If you really want to make sure your codebase can handle multithreading, you may want to do more explicit multithreaded testing. However, this approach is good for older codebases that deal with a lot of global state.
+
+So, let's give it a shot! As already mentioned, the codebase I focused on was NumPy. This project was my first major foray into open-source development, so when I saw I would be working with NumPy, I honestly was a little intimidated! NumPy is such a fundamental and widely used library, and I came into this not knowing a lot about how to contribute to something like that! Of course, I had no reason to feel nervous. The NumPy community is wonderful, and I enjoyed my time working with them, lurking around the community meetings, and learning what I could.
+
+Like any other big software project, NumPy has a test suite. During NumPy's lifetime, its test suite followed best-practices, first being written in `unittest`, then `nose`, and now `pytest`, a popular Python unit testing framework. Due to NumPy's size and lifetime, its test suite is fairly large and contains some older code.
+
+### NumPy + Test Suite + pytest-run-parallel = ???
+
+Now we have the three major parts of the project: free-threaded Python, NumPy, and its test suite written with `pytest`. So, how do we put these pieces together? How can we get the test suite running under multiple threads? For this, we can employ the help of one last library: [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel), a `pytest` plugin developed by folks here at Quansight.
+
+`pytest-run-parallel` is useful for multi-threading stress testing, exposing thread-safety issues in the test suite. The approach `pytest-run-parallel` takes to accomplish this is to run each test multiple times in parallel, utilizing multiple threads.
+
+![How pytest-run-parallel handles running tests. Tests runs one-by-one, multiple times under multiple threads.](/posts/numpy-thread-safe-test-suite/pytest-run-parallel-diagram.png)
```

Comment:
I think it'd help if you used the term "thread pool" in this caption. Each test gets its own thread pool. So the test is run many times in parallel, but only with itself, not other tests in the test suite.